File: internal/catalog/catalog.go
package catalog

import (
	"sort"
	"strings"
)

// ReasoningEffort describes one reasoning level exposed to Grok Build.
type ReasoningEffort struct {
	Value       string `json:"value"`
	Label       string `json:"label"`
	Description string `json:"description"`
	Default     bool   `json:"default"`
}

// ReasoningCapability describes the reasoning levels supported by a model.
type ReasoningCapability struct {
	DefaultEffort string
	Efforts       []ReasoningEffort
}

// Model describes the small amount of model metadata the proxy needs for
// request shaping and Grok Build's model picker.
type Model struct {
	ID            string               `json:"id"`
	DisplayName   string               `json:"display_name,omitempty"`
	Description   string               `json:"description,omitempty"`
	ContextWindow int                  `json:"context_window,omitempty"`
	ResponsesLite bool                 `json:"-"`
	Reasoning     *ReasoningCapability `json:"-"`
}

func reasoningCapability(defaultEffort string) *ReasoningCapability {
	efforts := []ReasoningEffort{
		{Value: "low", Label: "Low", Description: "Faster responses with lighter reasoning."},
		{Value: "medium", Label: "Medium", Description: "Balanced reasoning for most tasks."},
		{Value: "high", Label: "High", Description: "Deeper reasoning for complex tasks."},
		{Value: "xhigh", Label: "Extra high", Description: "Maximum supported reasoning depth."},
	}
	for i := range efforts {
		efforts[i].Default = efforts[i].Value == defaultEffort
	}
	return &ReasoningCapability{DefaultEffort: defaultEffort, Efforts: efforts}
}

func cloneModel(model Model) Model {
	if model.Reasoning == nil {
		return model
	}
	capability := *model.Reasoning
	capability.Efforts = append([]ReasoningEffort(nil), capability.Efforts...)
	model.Reasoning = &capability
	return model
}

var knownModels = map[string]Model{
	"gpt-5.6-sol": {
		ID:            "gpt-5.6-sol",
		DisplayName:   "GPT-5.6 Sol",
		Description:   "Latest frontier agentic coding model.",
		ContextWindow: 372000,
		ResponsesLite: true,
		Reasoning:     reasoningCapability("low"),
	},
	"gpt-5.6-terra": {
		ID:            "gpt-5.6-terra",
		DisplayName:   "GPT-5.6 Terra",
		Description:   "Balanced agentic coding model for everyday work.",
		ContextWindow: 372000,
		ResponsesLite: true,
		Reasoning:     reasoningCapability("medium"),
	},
	"gpt-5.6-luna": {
		ID:            "gpt-5.6-luna",
		DisplayName:   "GPT-5.6 Luna",
		Description:   "Fast agentic coding model.",
		ContextWindow: 372000,
		ResponsesLite: true,
		Reasoning:     reasoningCapability("medium"),
	},
	"gpt-5.5": {
		ID:            "gpt-5.5",
		DisplayName:   "GPT-5.5",
		Description:   "Frontier model for complex coding and real-world work.",
		ContextWindow: 272000,
		Reasoning:     reasoningCapability("medium"),
	},
	"gpt-5.2": {
		ID:            "gpt-5.2",
		DisplayName:   "GPT-5.2",
		Description:   "Model optimized for professional work and long-running agents.",
		ContextWindow: 272000,
	},
}

var defaultIDs = []string{
	"gpt-5.6-sol",
	"gpt-5.6-terra",
	"gpt-5.6-luna",
	"gpt-5.5",
	"gpt-5.2",
}

type Catalog struct {
	models map[string]Model
	order  []string
}

// New builds a catalog from a comma-separated allow-list. An empty string uses
// the current built-in defaults. Unknown IDs are accepted as full Responses API
// models so users can try newly enabled account-specific models without a new
// proxy release.
func New(csv string) Catalog {
	ids := splitIDs(csv)
	if len(ids) == 0 {
		ids = append([]string(nil), defaultIDs...)
	}
	models := make(map[string]Model, len(ids))
	order := make([]string, 0, len(ids))
	for _, id := range ids {
		base, _ := NormalizeID(id)
		if _, exists := models[base]; exists {
			continue
		}
		model, ok := knownModels[base]
		if !ok {
			model = Model{
				ID:            base,
				DisplayName:   base,
				ContextWindow: 272000,
				ResponsesLite: strings.HasPrefix(base, "gpt-5.6-"),
			}
		}
		models[base] = cloneModel(model)
		order = append(order, base)
	}
	return Catalog{models: models, order: order}
}

func splitIDs(csv string) []string {
	seen := map[string]struct{}{}
	var ids []string
	for _, value := range strings.Split(csv, ",") {
		id := strings.TrimSpace(value)
		if id == "" {
			continue
		}
		if _, ok := seen[id]; ok {
			continue
		}
		seen[id] = struct{}{}
		ids = append(ids, id)
	}
	return ids
}

// NormalizeID resolves the proxy's -fast alias to the upstream model ID.
func NormalizeID(id string) (base string, fast bool) {
	id = strings.TrimSpace(id)
	if strings.HasSuffix(id, "-fast") {
		candidate := strings.TrimSuffix(id, "-fast")
		if candidate != "" {
			return candidate, true
		}
	}
	return id, false
}

func (c Catalog) Lookup(id string) (Model, bool) {
	base, _ := NormalizeID(id)
	model, ok := c.models[base]
	if ok {
		return cloneModel(model), true
	}
	// A mapping target may be a built-in Codex model that the user chose not to
	// advertise via --models. Keep its authoritative request-shape metadata.
	if model, ok := knownModels[base]; ok {
		return cloneModel(model), true
	}
	// Unknown account-specific models are allowed through. Infer the only wire
	// distinction currently needed by the proxy from the model family.
	if base != "" {
		return Model{
			ID:            base,
			DisplayName:   base,
			ContextWindow: 272000,
			ResponsesLite: strings.HasPrefix(base, "gpt-5.6-"),
		}, false
	}
	return Model{}, false
}

func (c Catalog) Models() []Model {
	models := make([]Model, 0, len(c.order))
	for _, id := range c.order {
		models = append(models, cloneModel(c.models[id]))
	}
	return models
}

func (c Catalog) IDs() []string {
	ids := append([]string(nil), c.order...)
	return ids
}

// SortedKnownIDs is primarily useful for diagnostics and tests.
func SortedKnownIDs() []string {
	ids := make([]string, 0, len(knownModels))
	for id := range knownModels {
		ids = append(ids, id)
	}
	sort.Strings(ids)
	return ids
}

File: src/catalog.rs
use serde::Serialize;
use std::collections::{HashMap, HashSet};

const FAST_SUFFIX: &str = "-fast";

#[derive(Clone, Debug, Serialize, PartialEq, Eq)]
pub struct ReasoningEffort {
    pub value: String,
    pub label: String,
    pub description: String,
    pub default: bool,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct ReasoningCapability {
    pub default_effort: String,
    pub efforts: Vec<ReasoningEffort>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Model {
    pub id: String,
    pub display_name: String,
    pub description: String,
    pub context_window: u64,
    pub responses_lite: bool,
    pub reasoning: Option<ReasoningCapability>,
}

fn reasoning(default: &str) -> ReasoningCapability {
    let values = [
        ("low", "Low", "Faster responses with lighter reasoning."),
        ("medium", "Medium", "Balanced reasoning for most tasks."),
        ("high", "High", "Deeper reasoning for complex tasks."),
        ("xhigh", "Extra high", "Maximum supported reasoning depth."),
    ];
    ReasoningCapability {
        default_effort: default.into(),
        efforts: values
            .into_iter()
            .map(|(value, label, description)| ReasoningEffort {
                value: value.into(),
                label: label.into(),
                description: description.into(),
                default: value == default,
            })
            .collect(),
    }
}

fn known(id: &str) -> Option<Model> {
    let (name, description, context, lite, effort) = match id {
        "gpt-5.6-sol" => (
            "GPT-5.6 Sol",
            "Latest frontier agentic coding model.",
            372_000,
            true,
            Some("low"),
        ),
        "gpt-5.6-terra" => (
            "GPT-5.6 Terra",
            "Balanced agentic coding model for everyday work.",
            372_000,
            true,
            Some("medium"),
        ),
        "gpt-5.6-luna" => (
            "GPT-5.6 Luna",
            "Fast agentic coding model.",
            372_000,
            true,
            Some("medium"),
        ),
        "gpt-5.5" => (
            "GPT-5.5",
            "Frontier model for complex coding and real-world work.",
            272_000,
            false,
            Some("medium"),
        ),
        "gpt-5.2" => (
            "GPT-5.2",
            "Model optimized for professional work and long-running agents.",
            272_000,
            false,
            None,
        ),
        _ => return None,
    };
    Some(Model {
        id: id.into(),
        display_name: name.into(),
        description: description.into(),
        context_window: context,
        responses_lite: lite,
        reasoning: effort.map(reasoning),
    })
}

pub fn normalize_id(id: &str) -> (String, bool) {
    let id = id.trim();
    if let Some(base) = id.strip_suffix(FAST_SUFFIX).filter(|s| !s.is_empty()) {
        (base.into(), true)
    } else {
        (id.into(), false)
    }
}

#[derive(Clone, Debug)]
pub struct Catalog {
    models: HashMap<String, Model>,
    order: Vec<String>,
}

impl Catalog {
    pub fn new(csv: &str) -> Self {
        let defaults = "gpt-5.6-sol,gpt-5.6-terra,gpt-5.6-luna,gpt-5.5,gpt-5.2";
        let source = if csv.trim().is_empty() { defaults } else { csv };
        let mut seen = HashSet::new();
        let mut models = HashMap::new();
        let mut order = Vec::new();
        for raw in source.split(',') {
            let (id, _) = normalize_id(raw);
            if id.is_empty() || !seen.insert(id.clone()) {
                continue;
            }
            let model = known(&id).unwrap_or_else(|| Model {
                id: id.clone(),
                display_name: id.clone(),
                description: String::new(),
                context_window: 272_000,
                responses_lite: id.starts_with("gpt-5.6-"),
                reasoning: None,
            });
            models.insert(id.clone(), model);
            order.push(id);
        }
        Self { models, order }
    }

    pub fn lookup(&self, id: &str) -> (Model, bool) {
        let (base, _) = normalize_id(id);
        if let Some(model) = self.models.get(&base) {
            return (model.clone(), true);
        }
        if let Some(model) = known(&base) {
            return (model, true);
        }
        (
            Model {
                id: base.clone(),
                display_name: base.clone(),
                description: String::new(),
                context_window: 272_000,
                responses_lite: base.starts_with("gpt-5.6-"),
                reasoning: None,
            },
            false,
        )
    }
    pub fn models(&self) -> Vec<Model> {
        self.order
            .iter()
            .filter_map(|id| self.models.get(id).cloned())
            .collect()
    }
    pub fn ids(&self) -> Vec<String> {
        self.order.clone()
    }
}

impl Default for Catalog {
    fn default() -> Self {
        Self::new("")
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn catalog_defaults_and_fast_normalization() {
        let c = Catalog::default();
        assert_eq!(c.ids()[0], "gpt-5.6-sol");
        assert_eq!(
            normalize_id(" gpt-5.6-sol-fast "),
            ("gpt-5.6-sol".into(), true)
        );
        assert!(c.lookup("new-model").0.display_name == "new-model");
    }
}
